Packaging

austinbeauch opened the pull request:
About to start working on some tests and just wanted to see if I was on the right track for packaging the project. The import statements were a little weird and I had to change a couple lines so definitely check to see if it's correct, functionality should remain the same.

austinbeauch committed:
"Create package from existing scripts"

austinbeauch committed:
Add .idea directory to gitignore

austinbeauch commented:
Actually the imports are still a problem. Still working on it.

ijiraq committed:
made import absolute instead of relative so they work.

austinbeauch committed:
Merge pull request #1 from ijiraq/master

made import absolute instead of relative so they work.

austinbeauch committed:
Remove absolute_import statement, change module imports

austinbeauch committed:
Add test and header files for testing pv/sip transformations

austinbeauch commented on sip_tpv/tests/test_transforms.py:
```diff
@@ -0,0 +1,52 @@
+            -25.8095954324 -25.8095954322
+            19.1907457483 19.1907457481
+            """
+            assert q == k
```
These values don't come out 100% equal. Either I did something with the calls wrong or there's supposed to be some error, in which I can change it to be an assert almost equal equivalent so it passes.

austinbeauch commented on sip_tpv/pv_to_sip.py:
```diff
@@ -34,11 +34,12 @@
+                                sym_sipexprs,
+                                real_tpvexprs,
+                                add_sip_keywords,
+                                remove_pv_keywords)
```
Just changed the spacing here so PyCharm would stop getting mad at me.

stargaser commented on sip_tpv/tests/test_transforms.py:
```diff
@@ -0,0 +1,52 @@
+
+    for i, j in map(None, world2, world1):
+        for q, k in map(None, i, j):
+            print q, k
```
Please change to ``print(q, k)`` and add ``from __future__ import print_function`` at the top, to make it compatible with both Python 2 and 3.

austinbeauch replied:
I would normally take that out once the tests are working, for now I'll add in the change though.

stargaser commented on sip_tpv/tests/test_transforms.py:
```diff
@@ -0,0 +1,52 @@
+from astropy.io import fits
```
Add ``from __future__ import print_function`` at the top.

stargaser commented:
I tried a `python setup.py install` into a Python 3 conda environment -- it appeared to install to `sip_2pv` in the site-packages directory, but then also referred to `sip-2pv` which would not be a valid package name. I haven't yet figured out where this is being set...

stargaser commented:
Recommend to populate the ``sip_tpv/__init__.py`` file:

```
from .sip_to_pv import sip_to_pv
from .pv_to_sip import pv_to_sip
import pkg_resources

__version__ = pkg_resources.get_distribution("sip_tpv").version
```

stargaser commented on setup.py:
```diff
@@ -0,0 +1,22 @@
+from setuptools import setup
+
+
+setup(name='sip_2pv',
```
I would change the name to ``sip_tpv``. Then I find I can import it. I don't know myself what happens when the name and the package are different.

stargaser commented on sip_tpv/pvsiputils.py:
```diff
@@ -1,4 +1,4 @@
-from __future__ import print_function, absolute_import, division
+from __future__ import print_function, division
```
You can add ``absolute_import`` back in.

stargaser commented on sip_tpv/sip_to_pv.py:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python
 
-from __future__ import print_function, absolute_import, division
+from __future__ import print_function, division
```
can add ``absolute_import`` back in

stargaser commented on sip_tpv/sip_to_pv.py:
```diff
@@ -34,11 +34,11 @@
-                        real_sipexprs,
-                        add_pv_keywords,
-                        remove_sip_keywords)
+from sip_tpv.pvsiputils import (sym_tpvexprs,
```
can change to ``from .pvsiputils import (sym_tpvexprs,``

stargaser commented on sip_tpv/pv_to_sip.py:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python
 
-from __future__ import print_function, absolute_import, division
+from __future__ import print_function, division
```
Can add the ``absolute_import`` back in.

stargaser commented on sip_tpv/pv_to_sip.py:
```diff
@@ -34,11 +34,12 @@
-                       real_tpvexprs,
-                       add_sip_keywords,
-                       remove_pv_keywords)
+from sip_tpv.pvsiputils import (get_pv_keywords,
```
Can use ``from .pvsiputils import (get_pv_keywords,``

stargaser commented on sip_tpv/pvsiputils.py:
```diff
@@ -30,7 +30,7 @@
 from sympy import symbols, Matrix, poly
 import numpy as np
-import reverse
+from sip_tpv import reverse
```
Can change to ``from .reverse import fitreverse`` and change usage later down in the file.

stargaser replied:
Having trouble leaving a line comment.  Change line 472 from ``reverse.fitreverse`` to simply ``fitreverse``.

austinbeauch commented:
I accidentally put the same in __init__.py as sip_2pv, does python change underscores into dashes? I'll change the name to sip_tpv and try to install it. Same thing was happening to me in a python 2 conda environment.

stargaser commented:
I think the dashes are okay -- if name is set to sip_tpv then the package imports as sip_tpv.

stargaser reviewed (changes requested):
With the line comment changes, I can do:

- (in Python 3 env) ``python setup.py install`` from the top-level.
- ``pytest`` from the top-level (it auto-discovers the tests)

The tests don't run successfully for me yet. I'll submit the review for now.

I haven't tried Python 2 yet. ``python setup.py develop`` might work from the top-level?

austinbeauch committed:
Change package name to 'sip_tpv' instead of 'sip_2pv'

Changed back the import statements, using the form from package.module import methods did not work from commant line.

austinbeauch commented:
Okay I can hop into a python prompt and `>>> import sip_tpv` works.

austinbeauch commented:
Changes you requested just showed up for me, I'll get on those now.

austinbeauch committed:
Add absolute_import back in and fix import statements

stargaser commented:
Many thanks for the fast updates! Looks good. I'm afraid things got busy for me and I'll have to be away from keyboard again.

I have some updates to the test_transforms.py.

Can make for the import
```
from ..sip_to_pv import sip_to_pv
from ..pv_to_sip import pv_to_sip
```

can ``import os``, and include
```
dir_name = os.path.split(__file__)[0]
```

And then when referencing the file, use
```
fits.Header.fromtextfile(os.path.join(dir_name,'data/IRAC_3.6um_sip.txt'))
```

And then the test data can be found when running pytest from other directories.

For the actual test, I get an exception
```
TypeError: 'NoneType' object is not callable
```
at the ``for`` loop.

I think we could use ``numpy.testing.assert_almost_equals`` and other methods, rather than the loop. Would you have a chance to try it out?

Thanks for all this!

austinbeauch committed:
Use numpy.testing for testing values

Update import statements to be relative. Change variable names for more clarity. Add "Test" prints to determine which tests are running.

austinbeauch commented:
It is beyond me, but for whatever reason only one of the test methods would run with their original names so I added some quick print statements and changed a method name which got them to run. The tests ran successfully inside PyCharm with nosetests, although converting pv->sip only had accuracy of 3 decimal points - whereas sip->pv had 100% accuracy. Also I used the all_pix2world method each time for changing them into coordinates because that's what happened to work and unfortunately I'm not too familiar with all of astropy's methods.

stargaser commented on sip_tpv/tests/test_transforms.py:
```diff
@@ -1,54 +1,43 @@
-    pvheader = fits.Header.fromtextfile('data/PTF_r_chip01_tpv.txt')
-    header = pvheader.copy()
-    pv_to_sip(header)
+def test_():
```
hmm, can't see why test_pv2sip wouldn't work?

stargaser commented on sip_tpv/tests/test_transforms.py:
```diff
@@ -1,54 +1,43 @@
-    header = pvheader.copy()
-    pv_to_sip(header)
+def test_():
+    pv_header = fits.Header.fromtextfile(os.path.join(dir_name, 'data/IRAC_3.6um_sip.txt'))
```
Ah, this could be a problem since that header has SIP and not TPV. Did the other header not work for some reason?

austinbeauch replied:
Whoops, used the wrong file! Now the results are exactly equal. The method name still won't work when run as test_pv2sip though. I'll ask around today and see if anyone knows what's going on.

austinbeauch replied:
Well this is bizarre... sometimes when I run the test both methods will run, and other times just the one will run. They're all working though, which I suppose is the most important part.

stargaser commented on sip_tpv/tests/test_transforms.py:
```diff
@@ -1,54 +1,43 @@
+    wsip = WCS(pv_header)
+    wtpv = WCS(control_header)
 
     world1 = wsip.all_pix2world([[1, 1]], 1)
```
Here and elsewhere, more coordinates could be added -- the first argument to ``all_pix2world`` can be an Nx2 array if I recall correctly. So you could do
```
import numpy as np
pixarr = np.array([[1,1], [400,1], [400,400], [1,400])
world1 = wsip.all_pix2world(pixarr, 1)
```

stargaser replied:
To follow this up...one could pull out NAXIS1 and NAXIS2 from the header, and make a sparse grid of pixel coordinates for the test. Better than hardcoding 400.

austinbeauch replied:
Is this what you mean?

```
naxis1 = pv_header['NAXIS1']
naxis2 = pv_header['NAXIS2']
pixarr = np.array([[1, 1], [naxis1, 1], [naxis1, naxis2], [1, naxis2]])
```

stargaser replied:
That does the corners. Here is one way to do a 10x10 grid to get 100 points to test:

```
x = np.linspace(1, naxis1, 10)
y = np.linspace(1, naxis2, 10)
xx, yy = np.meshgrid(x,y)
pixargs = np.vstack([xx.reshape(-1), yy.reshape(-1)]).T
```

stargaser commented on sip_tpv/tests/test_transforms.py:
```diff
@@ -1,54 +1,43 @@
     world1 = wsip.all_pix2world([[1, 1]], 1)
     world2 = wtpv.all_pix2world([[1, 1]], 1)
+    print("Test2")
+    npt.assert_array_almost_equal(world1, world2, 3)
```
Converting sip to pv, in the pixel-to-sky direction, I would expect it would be good to 6 digits.  Might be caused by issues above...or maybe indicates a problem in the conversion.

Also I just took a look at the docs for ``assert_almost_equal`` and I see they recommend using some other functions...sorry I didn't notice when glancing at it earlier today.

austinbeauch replied:
Since they're now completely equal, will `assert_equal` work?

stargaser replied:
``assert_equal`` might be worth a try...I am not sure how these methods handle machine precision.

stargaser reviewed (commented):
Installation works for me in both py2 and py3 environments, and pytest runs the tests successfully. This is great!

I think it would be good to try the TPV header in the second test as maybe that will fix the agreement to only 3 decimal places.

austinbeauch committed:
Change PV test to use correct header file

Create larger pixel arrays for unit conversion tests.

austinbeauch committed:
Create pixel grid from NAXIS keywords

austinbeauch committed:
Create 10x10 grid from NAXIS keywords

austinbeauch commented:
Ah I figured out the test problems, it was only running the test method where my cursor was!

stargaser commented:
Glad to hear the test naming is working now. I guess it must be a PyCharm feature that was causing confusion.

If you are willing, it would also be good to test with the `wcs.all_world2pix` methods. These could be added to your existing tests. It might be that `npt.all_equal` will be too stringent, because the world2pix direction is either an iterative solution inverting the forward polynomials; or in the case of SIP can sometimes be the "reverse polynomials".

One small point on your "Test 3". I think you need to do something like `wtpv2 = WCS(pv_header)` and then test with that WCS object. I don't think the original `wsip` object will update automatically when the header object is updated.

stargaser commented:
The testing shows that the way I coded the conversion routines to modify the header, is a problem. It would be better for the conversion functions to make a copy of the header, modify it and return it, instead of returning `None`. Then the headers could keep their appropriate names, and we could use `sip_header = pv_to_sip(pv_header)`. That would be easier to read.

Maybe that should be saved for another pull request. I think that after adding some `world2pix` tests, we could declare victory and merge this one. What do you think?

austinbeauch commented:
Yes, I think the sooner we can get it confirmed to be working and merged the better! That way we can get it published on PyPi.
I'll get to work on the other tests/changes right after a quick lunch break, had some meetings this morning.

austinbeauch commented:
I don't think I'm using all_world2pix correctly... Taking the example from the docstrings I'm trying to do `ra, dec = wsip.all_pix2world([1,2,3], [1,1,1], 1)` but that keeps throwing an error 

`NoConvergence: 'WCS.all_world2pix' failed to converge to the requested accuracy.
After 2 iterations, the solution is diverging at least for one input point.`

stargaser commented:
The inputs to `all_pix2world` and `all_world2pix` need to have shape Nx2. 

You could feed in `world1` Etc.

stargaser commented:
Also I think if your world coordinates lie far from the image, `all_world2pix` will show these kind of convergence problems. I know there was an issue in Astropy before that I was consulted about

austinbeauch committed:
Test round trip for pix2world and world2pix

austinbeauch commented:
JJ (my supervisor) sat down and helped run me through some things, said the tests should be good now. We checked the docs for `assert_almost_equal` and came to the conclusion that it's sufficient for these purposes, mainly because the other methods the docs suggested were not working properly. I think they have use cases outside of these tests.

austinbeauch commented:
Also, do you want me to go through and update the formatting so it all follows PEP 8 convention? I would be more than happy to  update the main files.

stargaser commented:
Yes, please do the PEP8 formatting. That will be great!

stargaser reviewed (approved):
Terrific! I like the roundtripping checks.

austinbeauch commented:
I'll get the PEP8 done within the hour (because in an hour it's 5 and I get to go home).

austinbeauch committed:
Update to follow PEP8 styling conventions

austinbeauch commented:
There are a few variables that don't get used in a couple methods but I just completely left all that alone. Test still works so I think its good!

stargaser commented:
Merged. Thank you @austinbeauch ! 🎆

austinbeauch commented:
@stargaser do you plan on publishing this to PyPi?

stargaser commented:
yes, but I think there is some package cleanup to do, at least for the README. Also I am not sure if everything needed in `setup.py` is there. Let me review my notes on publishing and I'll try publishing it to the test server.

stargaser commented:
@austinbeauch ok, actually it seems to have worked to the test server. If you want you can try
```
pip install --verbose --index-url https://testpypi.python.org/pypi sip_tpv
```

austinbeauch commented:
Awesome! I'm fine with just running the local install, just needed to make sure you had plans of publishing it so when our project is done and other people use it they can just download this package from PyPi.

stargaser commented:
OK. I went ahead and put this version on PyPI.org since it seems to work. I'll leave fix-ups to another version. Also I'd like to change the behavior of modifying headers in-place, to leaving the input header alone and returning a new header. That will be an API change so it would be for a 2.0 version.
